File: behavioral/dart/observer_structured/youtube_channel.dart
import 'interfaces.dart';

class YouTubeChannel extends Subject {
  final String _channelName;
  
  YouTubeChannel(this._channelName);
  
  void uploadVideo(String videoTitle) {
    print('\n$_channelName uploaded: $videoTitle');
    notify('New video: $videoTitle');
  }
  
  void goLive(String streamTitle) {
    print('\n$_channelName is live: $streamTitle');
    notify('Live stream started: $streamTitle');
  }
}

class Subscriber implements Observer {
  final String _name;
  
  Subscriber(this._name);
  
  @override
  void update(String message) {
    print('  $_name received notification: $message');
  }
}

class NotificationService implements Observer {
  @override
  void update(String message) {
    print('  [NotificationService] Push notification sent: $message');
  }
}

File: behavioral/typescript/observer_structured/youtube_channel.ts
import { Observer, Subject } from './interfaces';

export class YouTubeChannel extends Subject {
  private channelName: string;
  
  constructor(name: string) {
    super();
    this.channelName = name;
  }
  
  uploadVideo(videoTitle: string): void {
    console.log(`\n${this.channelName} uploaded: ${videoTitle}`);
    this.notify(`New video: ${videoTitle}`);
  }
  
  goLive(streamTitle: string): void {
    console.log(`\n${this.channelName} is live: ${streamTitle}`);
    this.notify(`Live stream started: ${streamTitle}`);
  }
}

export class Subscriber implements Observer {
  private name: string;
  
  constructor(name: string) {
    this.name = name;
  }
  
  update(message: string): void {
    console.log(`  ${this.name} received notification: ${message}`);
  }
}

export class NotificationService implements Observer {
  update(message: string): void {
    console.log(`  [NotificationService] Push notification sent: ${message}`);
  }
}
